Object service

owlishDeveloper opened the pull request:
Note: I didn't entirely understand the mechanics/intentions of the API in the original RFC, so I made my own (maybe re-invented the wheel, really).

Feel free to ask questions, critique, etc.

djmitche commented on rfcs/Create-object-service.md:
```diff
@@ -0,0 +1,53 @@
+Taskcluster tasks often produce artifacts during the course of being run.
+Examples include browser distribution archives, json metadata and log files.
+We currently store artifacts in a single region of S3 and split management of
+artifacts between the Queue and allegedly Cloud Mirror.  
```
"allegedly" -> "until November 2019"?

owlishDeveloper replied:
I haven't seen artifacts marked "cloud mirror" later than 2015 and 2016, but if it actually worked until November 2019 then yes

djmitche commented on rfcs/Create-object-service.md:
```diff
@@ -0,0 +1,53 @@
+
+The service is to be implemented in such a way that it could be as easily as possible switched from Azure to Postgres.
+The relational database for the service should be designed as well (in the form of an entity relationship diagram as part
+of the documentation, if Postgres support will not be deployed by the time of the prototype development).
```
This timing might work out!  We are hoping to land tc-lib-postgres before the all-hands, at which time I think it would be OK to start using it for new development while we continue to work on migration of the existing services.

owlishDeveloper commented:
> I can't see the swagger:
> 
> ![image](https://user-images.githubusercontent.com/28673/72638763-20395e00-3932-11ea-8d7b-a6a9a91d5385.png)
> 
My bad - it was private. Can you check again? It should be public now

> Overall this looks good! I like that it's high-level, but provides a framework for developing the details. I'd like to know a little more about what "efficient" means: saving bandwidth costs? download times? costs for additional storage?

"efficient" means cost-efficient and also time-efficient. It would be nice to find the cheapest ways of doing things (in terms of money - any money, storage or bandwidth) and in terms of how long things take, or find some reasonable compromise.

djmitche commented:
I can see it now -- thanks!

ccooper commented on rfcs/Create-object-service.md:
```diff
@@ -0,0 +1,53 @@
+The relational database for the service should be designed as well (in the form of an entity relationship diagram as part
+of the documentation, if Postgres support will not be deployed by the time of the prototype development).
+
+The service should be compatible with the current implementation (the object storage code in the Queue service), 
+and require as little changes to the clients and workers as possible (preferably, none at all).
```
It might be worth following a similar approach to what dustin/hassan are doing with azure support, i.e. breaking out the object-handling portion of the queue into a standalone service without making any changes at all, and then iterating on that new service.

owlishDeveloper replied:
> It might be worth following a similar approach to what dustin/hassan are doing with azure support, i.e. breaking out the object-handling portion of the queue into a standalone service without making any changes at all, and then iterating on that new service.

Makes sense to me! I'll start hacking on this as a part of my planning. I'll also time-box writing a little script that puts an object into S3 and gets a hash of the data back. I need to do a bit of hacking because it would help me to plan and to understand how much work this is. I'll time-box the script tho to prevent sliding into a rabbit hole. Separating the functionality first and making sure it works the same way (and defining "the same way") and then changing it makes sense to me as an overall plan.

ccooper commented on rfcs/Create-object-service.md:
```diff
@@ -0,0 +1,53 @@
+
+# Motivation
+
+* We want to pay less, so we want to be able to store things in different regions and clouds
```
I think we'll probably continue storing things in S3 if for no other reason than inertia, but being able to manage ingress from the other providers, and temporarily cache artifacts in a given provider where they might be needed repeatedly (e.g. builds for testing) are both standout use cases.

You should consider adding the most important use cases that you're using to justify this new service.

ccooper commented on rfcs/Create-object-service.md:
```diff
@@ -0,0 +1,53 @@
+
+* We want to pay less, so we want to be able to store things in different regions and clouds
+* We want the Queue to be easier to understand, develop and maintain
+* We want object management to be easier to understand, develop and maintain
```
I would put "We want object management to be easier to understand, develop and maintain" last in the list. This would bring the two Queue-related items together and give the list a stronger finish.

ccooper commented on rfcs/Create-object-service.md:
```diff
@@ -0,0 +1,53 @@
+The service should allow for flexibility in choosing artifact/object encryption methods and algorithms.
+
+Implementation of any cross-region and/or cross-cloud functionality must be justified as far as money and time cost goes.
+The service must be efficient.
```
...as measured how? Cost? Latency? Redundancy? ...?

milescrabill commented on rfcs/Create-object-service.md:
```diff
@@ -0,0 +1,82 @@
+### Updating
+No such endpoint. We don't want the objects to be updatable for security reasons.
+
+###Deletion
```
nit: add a space here

owlishDeveloper replied:
😛

owlishDeveloper commented:
>     * the swagger definition doesn't have the `delete` operation listed
The swagger definition is no longer relevant, I removed it from the document. I added DELETE as I was copying the API definition from swagger..... I realized that since saving cost is one of the top priorities now, implementing this operation makes more sense than it did when the original discussions took place. So I added it.

>     how does the object service handle object signatures when objects are uploaded directly to cloud providers?
I imagine object signatures would be created once the object has been uploaded (can be done as a step 3 in the creation process). The three cloud providers I looked at (AWS, GCP, Azure) do offer a server-side encryption with customer-managed keys (as well as the cloud provider managed keys). At the moment, it makes sense to me to utilize that functionality. Does this answer your question?

djmitche commented:
Good point about DELETE.  Will it be the queue's responsibility to remember about artifacts and call the DELETE method on the corresponding object, or will the Object Service have a way to support automatically expiring objects?  At the moment, the queue uses the first approach (and it is careful to write to its artifacts table before creating the object, so that if something crashes, it doesn't leave an orphan object).

sciurus commented:
I don't see any mention of CDNs in the RFC, but I assume the object service will continue giving out CDN urls instead of direct S3 (or equivalent) URLs in the situations where this is done today.

owlishDeveloper commented:
>  Will it be the queue's responsibility to remember about artifacts and call the DELETE method on the corresponding object, or will the Object Service have a way to support automatically expiring objects?

I wonder of we can have both? Support expiring objects through the bucket life-cycle settings (I'm not sure about this functionality in GCP and Azure, would need to look that up), _and_ have an endpoint in object service that can either be called by a person (curl etc) or be put into a script like a cron job. The state of this will be saved in database, of course

owlishDeveloper commented:
> I don't see any mention of CDNs in the RFC, but I assume the object service will continue giving out CDN urls instead of direct S3 (or equivalent) URLs in the situations where this is done today.

My bad! I'll add a mention, that definitely should be in there, thank you. And to answer your question - yes, I think so.

> > Will it be the queue's responsibility to remember about artifacts and call the DELETE method on the corresponding object, or will the Object Service have a way to support automatically expiring objects?
> 
> an endpoint in object service that can either be called by a person (curl etc) or be put into a script like a cron job

...or be automatic based on the expiry setting just as queue does. I think if we have an API endpoint, we will have all of these options available to us to use

owlishDeveloper commented:
@djmitche but we _don't_ want to have an endpoint, but just some unexposed functionality within object service to periodically delete objects - perfectly fine with me as well. I can add this to the RFC and remove the `DELETE` endpoint then

djmitche commented:
I think what you suggested makes sense: be flexible in how artifacts are deleted, allowing both an automatic expiry and an explicit delete.  That mirrors what we do with other things like secrets.  If that expiry can happen "automatically" by some configuration of the cloud provider, all the better!
